File: pysrc/coating.cpp
#include "coating.h"
#include <pybind11/pybind11.h>

namespace py = pybind11;
using namespace pybind11::literals;


namespace batoid {
    void pyExportCoating(py::module& m) {
        py::class_<CoatingHandle, std::shared_ptr<CoatingHandle>>(m, "CPPCoating")
            .def("getCoefs", [](const CoatingHandle& handle, double wavelength, double cosIncidenceAngle){
                double reflect, transmit;
                handle.getHostPtr()->getCoefs(wavelength, cosIncidenceAngle, reflect, transmit);
                return py::make_tuple(reflect, transmit);
            })
            .def("getReflect", [](const CoatingHandle& handle, double wavelength, double cosIncidenceAngle){
                return handle.getHostPtr()->getReflect(wavelength, cosIncidenceAngle);
            })
            .def("getTransmit", [](const CoatingHandle& handle, double wavelength, double cosIncidenceAngle){
                return handle.getHostPtr()->getTransmit(wavelength, cosIncidenceAngle);
            });

        py::class_<SimpleCoatingHandle, std::shared_ptr<SimpleCoatingHandle>, CoatingHandle>(m, "CPPSimpleCoating")
            .def(py::init<double,double>(), "reflectivity"_a, "transmissivity"_a);
    }
}

File: batoid/coating.py
from batoid.trace import reflect
from . import _batoid


class Coating:
    """Class to control ray reflection/transmission at an `Interface`.

    Coatings can be used to split a ray into reflected/refracted components
    using `Surface.rSplit`, or control the transmission or reflection
    efficiency using `Surface.refract` or `Surface.reflect` (or variations
    thereof).

    In general, the reflection and transmission coefficients may depend on both
    wavelength and the cosine of the incidence angle, which is the angle
    between the incoming ray and the surface normal.
    """
    def getCoefs(self, wavelength, cosIncidenceAngle):
        """Return reflection and transmission coefficients.

        Parameters
        ----------
        wavelength : float
            Vacuum wavelength in meters.
        cosIncidenceAngle : float
            Cosine of the incidence angle.

        Returns
        -------
        reflect : float
        transmit : float
        """
        return self._coating.getCoefs(wavelength, cosIncidenceAngle)

    def getReflect(self, wavelength, cosIncidenceAngle):
        """Return reflection coefficient.

        Parameters
        ----------
        wavelength : float
            Vacuum wavelength in meters.
        cosIncidenceAngle : float
            Cosine of the incidence angle.

        Returns
        -------
        reflect : float
        """
        return self._coating.getReflect(wavelength, cosIncidenceAngle)

    def getTransmit(self, wavelength, cosIncidenceAngle):
        """Return transmission coefficient.

        Parameters
        ----------
        wavelength : float
            Vacuum wavelength in meters.
        cosIncidenceAngle : float
            Cosine of the incidence angle.

        Returns
        -------
        transmit : float
        """
        return self._coating.getTransmit(wavelength, cosIncidenceAngle)

    def __ne__(self, rhs):
        return not (self == rhs)


class SimpleCoating(Coating):
    """Coating with reflectivity and transmissivity that are both constant with
    wavelength and incidence angle.

    Parameters
    ----------
    reflectivity : float
        Reflection coefficient
    transmissivity : float
        Transmission coefficient
    """
    def __init__(self, reflectivity, transmissivity):
        self.reflectivity = reflectivity
        self.transmissivity = transmissivity
        self._coating = _batoid.CPPSimpleCoating(
            reflectivity, transmissivity
        )

    def __eq__(self, rhs):
        return (
            isinstance(rhs, SimpleCoating)
            and self.reflectivity == rhs.reflectivity
            and self.transmissivity == rhs.transmissivity
        )

    def __getstate__(self):
        return self.reflectivity, self.transmissivity

    def __setstate__(self, args):
        self.__init__(*args)

    def __hash__(self):
        return hash(("SimpleCoating", self.reflectivity, self.transmissivity))

    def __repr__(self):
        return f"SimpleCoating({self.reflectivity}, {self.transmissivity})"
